catkit/hardware/boston/BostonDmController.py

```diff
@@ -109,64 +113,96 @@ def apply_shape_to_both(self, dm1_shape, dm2_shape,
-                                  flat_map=flat_map,
-                                  bias=bias,
-                                  as_voltage_percentage=as_voltage_percentage,
-                                  as_volts=as_volts,
-                                  sin_specification=sin_specification)
-
-        if not isinstance(dm2_shape, DmCommand):
-            dm2_shape = DmCommand(data=dm2_shape,
-                                  dm_num=2,
-                                  flat_map=flat_map,
-                                  bias=bias,
-                                  as_voltage_percentage=as_voltage_percentage,
-                                  as_volts=as_volts,
-                                  sin_specification=sin_specification)
-
-        # Ensure that the correct dm_num is set.
-        dm1_shape.dm_num = 1
-        dm2_shape.dm_num = 2
-
-        if output_path is not None:
-            dm1_shape.export_fits(output_path)
-            dm2_shape.export_fits(output_path)
-
-        # Use DmCommand class to format the commands correctly (with zeros for other DM).
-        dm1_command = dm1_shape.to_dm_command()
-        dm2_command = dm2_shape.to_dm_command()
-
-        # Add both arrays together (first half and second half) and send to DM.
-        full_command = dm1_command + dm2_command
-        try:
-            self.send_data(full_command)
-        except Exception:
-            # We shouldn't guarantee the state of the DM.
-            self._clear_state()
-            raise
-        else:
-            # Update both dm_command class attributes.
-            self.dm1_command = dm1_command
-            self.dm2_command = dm2_command
-            self.dm1_command_object = dm1_shape
-            self.dm2_command_object = dm2_shape
+        if do_logging:
+            if channel is None:
+                self.log.info("Applying shape to both DMs")
+            else:
+                self.log.info(f'Applying shape to both DMs in channel {channel}.')
+
+        with self.lock:
+            if channel is not None:
+                if isinstance(dm1_shape, DmCommand) or isinstance(dm2_shape, DmCommand):
+                    # DmCommand objects cannot be added together, yet.
+                    raise ValueError('DM shapes cannot be DmCommands when using channels.')
+
+                # Check if dm{1,2}_shape is 2D, then convert to 1D.
+                if dm1_shape.ndim == 2:
+                    dm1_shape = convert_dm_image_to_command(dm1_shape)
+                if dm2_shape.ndim == 2:
+                    dm2_shape = convert_dm_image_to_command(dm2_shape)
+
+                self.channels[channel] = (dm1_shape, dm2_shape)
+
+                # Add contributions for each channel, and use that as the dm command.
+                dm1_shape = 0
+                dm2_shape = 0
+
+                for dm1, dm2 in self.channels.values():
+                    dm1_shape += dm1
+                    dm2_shape += dm2
```

Comment:
Comment added in the docstring of the function.

I'm now clearing the channels if `channel is None`. I'm also emitting a warning to the log, because this is extremely likely due to misuse of the channels API.